File: Client/BetterMusicPlayer/Behavior/Types/Device.swift
//
//  Device.swift
//  BetterMusicPlayer
//
//  Created by John Butler on 3/28/26.
//

struct Device {
    let id: String?
    let name: String
    
    let isActive: Bool
    let isPrivateSession: Bool
    let isRestricted: Bool
    let type: DeviceType
    let volume: Int?
    let supportsVolume: Bool
    
    init?(json: Dictionary<String, AnyObject>){
        guard
            let isActive = json["is_active"] as? Bool,
            let isPrivateSession = json["is_private_session"] as? Bool,
            let isRestricted = json["is_restricted"] as? Bool,
            let name = json["name"] as? String,
            let typeString = json["type"] as? String,
            let supportsVolume = json["supports_volume"] as? Bool
        else { return nil; }
        
        self.id = json["id"] as? String;
        self.name = name;
        self.isActive = isActive;
        self.isPrivateSession = isPrivateSession
        self.isRestricted = isRestricted;
        self.type = DeviceType(string: typeString);
        self.volume = json["volume_percentage"] as? Int;
        self.supportsVolume = supportsVolume;
    }
}

enum DeviceType {
    case Computer
    case SmartPhone
    case Speaker
    case Other
    
    init(string: String) {
        switch string {
        case "computer":
            self = .Computer
        case "smartphone":
            self = .SmartPhone
        case "speaker":
            self = .Speaker
        default:
            self = .Other
        }
    }
}

File: Server/src/core_types/device.rs
use crate::core_types::DeviceID;

#[derive(Debug, Clone, serde::Deserialize)]
#[allow(dead_code)]
pub struct Device {
    id: Option<DeviceID>,
    name: String,
    is_active: bool,
    is_private_session: bool,
    is_restricted: bool,
    #[serde(rename = "type")]
    device_type: DeviceType,
    #[serde(rename = "volume_percentage")]
    volume: Option<u8>,
    supports_volume: bool,
}

impl Device {
    pub fn id(&self) -> Option<&DeviceID> {
        self.id.as_ref()
    }
    pub fn is_active(&self) -> bool {
        self.is_active
    }
    pub fn name(&self) -> &str {
        &self.name
    }
}

impl PartialEq for Device {
    fn eq(&self, other: &Self) -> bool {
        self.id == other.id
    }
}

#[derive(Debug, Clone, Copy, serde::Deserialize, PartialEq, Eq)]
pub enum DeviceType {
    #[serde(rename = "computer")]
    Computer,
    #[serde(rename = "smartphone")]
    SmartPhone,
    #[serde(rename = "speaker")]
    Speaker,
    #[serde(other)]
    Other,
}
